stdlib/public/Darwin/Foundation/ContiguousBytes.swift

```diff
@@ -0,0 +1,100 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2018 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+//===--- ContiguousBytes --------------------------------------------------===//
+
+/// Indicates that the conforming type is a contiguous collection of raw bytes
+/// whose underlying storage is directly accessible by withUnsafeBytes.
+public protocol ContiguousBytes {
+    /// Calls the given closure with the contents of underlying storage.
+    ///
+    /// - note: Calling `withUnsafeBytes` multiple times does not guarantee that
+    ///         the same buffer pointer will be passed in every time.
+    /// - warning: The buffer argument to the body should not be stored or used
+    ///            outside of the lifetime of the call to the closure.
+    func withUnsafeBytes<R>(_ body: (UnsafeRawBufferPointer) throws -> R) rethrows -> R
+}
+
+//===--- Collection Conformances ------------------------------------------===//
+
+// FIXME: When possible, expand conformance to `where Element : Trivial`.
+extension Array : ContiguousBytes where Element == UInt8 { }
```

Comment:
given this will become ABI and can't be loosened later, are you _sure_ you want to add this? You can use `withContiguousStorageIfAvailable` instead in this case, right?

Reply:
(I don't mind adding it, just thinking you might want to reserve flexibility for yourself)

Reply:
We cannot use withContiguousStorageIfAvailable because this is being used as a protocol to identify it as a contiguous DataProtocol

Reply:
@itaiferber Right. And you can only have one conditional conformance. I can envision ABI-stable ways to loosen conditional conformances in the future, but we don't have any right now. cc @DougGregor 

@phausler but you have an init from a general sequence of UInt8, and so can have Array follow that path then fast-path into withContiguousStorageIfAvailable, no? Might amount to much the same thing.

Reply:
@airspeedswift Sure, the conditional conformance limitation I'm aware of. In the future, we're almost certainly going to need an ABI-stable way to loosen these anyway; even if not, we're willing to live with this for now.

We're also not using this _just_ in `init<S : Sequence>` — it's important for general consumers of `DataProtocol` who rely on `ContiguousBytes` for the `associatedtype` requirement.